pandas/core/groupby/generic.py

```diff
@@ -1309,17 +1309,21 @@ def _transform_general(self, func, *args, **kwargs):
         gen = self.grouper.get_iterator(obj, axis=self.axis)
         fast_path, slow_path = self._define_paths(func, *args, **kwargs)
 
+        path = None
         for name, group in gen:
             object.__setattr__(group, "name", name)
 
-            # Try slow path and fast path.
-            try:
-                path, res = self._choose_path(fast_path, slow_path, group)
-            except TypeError:
-                return self._transform_item_by_item(obj, fast_path)
-            except ValueError as err:
-                msg = "transform must return a scalar value for each group"
-                raise ValueError(msg) from err
+            if path is None:
```

Comment:
instead of doing this at each step in the loop, can we just do this outside the loop?

Reply:
Yes - updated PR to show what this would look like. Minor cleanups can be done to avoid the nested if statements in the newly added `process_result`, but it's a bit ugly. We're trading `if path is None` for a function call as I don't see another way to avoid code duplication.